tests/test_lora/test_gemini_lora.py

```diff
@@ -0,0 +1,196 @@
+        dist.barrier()
+
+        # The Lora checkpoint should be small in size
+        model_checkpoint_size_mb = os.path.getsize(os.path.join(lora_ckpt_path, "adapter_model.bin")) / (1024 * 1024)
+        optimizer_checkpoint_size_mb = os.path.getsize(optimizer_ckpt_path) / (1024 * 1024)
+        assert model_checkpoint_size_mb < 1 and optimizer_checkpoint_size_mb < 1
+
+        new_plugin = GeminiPlugin(
+            max_norm=1.0,
+            initial_scale=2**5,
+            master_weights=master_weights,
+            **placement_config,
+        )
+        new_booster = Booster(plugin=new_plugin)
+        model_load = new_booster.enable_lora(model_load, pretrained_dir=lora_ckpt_path)
+        optimizer_load = HybridAdam(model_load.parameters(), lr=0.001)
+        model_load, optimizer_load, _, _, _ = new_booster.boost(model_load, optimizer_load)
+
+        booster.load_optimizer(optimizer_load, optimizer_ckpt_path)
+
+        check_state_dict_equal(model_save.state_dict(only_rank_0=False), model_load.state_dict(only_rank_0=False))
+        check_state_dict_equal(
+            optimizer_save.state_dict(only_rank_0=False), optimizer_load.state_dict(only_rank_0=False)
+        )
+
+
+@clear_cache_before_run()
+def check_grad_accum(placement_config):
+    model_fn, data_gen_fn, output_transform_fn, loss_fn, _ = next(
+        iter(model_zoo.get_sub_registry("transformers_llama_for_casual_lm").values())
+    )
+
+    model = model_fn()
+    model.gradient_checkpointing_enable()
+
+    plugin = GeminiPlugin(
+        max_norm=1.0,
+        initial_scale=2**5,
+        enable_gradient_accumulation=True,
+        **placement_config,
+    )
+    booster = Booster(plugin=plugin)
+
+    lora_config = LoraConfig(task_type="CAUSAL_LM", r=8, lora_alpha=32, lora_dropout=0.1)
+    model = booster.enable_lora(model, lora_config=lora_config)
+    model_copy = copy.deepcopy(model)
+
+    optimizer = HybridAdam(model.parameters(), lr=0.001)
+    criterion = loss_fn
+
+    model, optimizer, criterion, _, _ = booster.boost(model, optimizer, criterion)
+
+    # Do forward and backward under grad accum setting.
+    accum_iter = 2
+    for i in range(2 * accum_iter):
+        data = data_gen_fn()
+        data = {
+            k: v.to("cuda") if torch.is_tensor(v) or "Tensor" in v.__class__.__name__ else v for k, v in data.items()
+        }
+        output = model(**data)
+        output = output_transform_fn(output)
+        loss = criterion(output) / accum_iter
+        booster.backward(loss, optimizer)
+
+        if (i + 1) % accum_iter == 0:
+            optimizer.clip_grad_by_norm(1.0)
+            optimizer.step()
+
+    # Check parameters
+    gemini_dict = model.state_dict(only_rank_0=False)
+    model_copy_dict = model_copy.state_dict()
+
+    for name in gemini_dict:
+        check_param_equality(
+            name, gemini_dict[name], model_copy_dict[name], modules_to_save=model.unwrap().modules_to_save
+        )
```

Comment:
Can you merge these function to single one?

Reply:
Done